colcon_sanitizer_reports/_sanitizer_section_part.py

```diff
@@ -0,0 +1,135 @@
+# Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from collections import ChainMap, defaultdict
+import re
+from typing import List, Optional, Tuple
+
+from colcon_sanitizer_reports._sanitizer_section_part_stack_trace import \
+    SanitizerSectionPartStackTrace
+
+
+_FIND_RELEVANT_STACK_TRACE_BEGIN_REGEXES_BY_ERROR_NAME = ChainMap({
+    # There are two relevant stack traces involved in a "data race" section part. Their headers
+    # match the following patterns.
+    'data race': (
+        re.compile(r'^\s+(Read|Write) of size \d+ at 0x[\da-f]+ .*$'),
+        re.compile(r'^\s+Previous (read|write) of size \d+ at 0x[\da-f]+ .*$'),
+    ),
+    # There is one relevant stack trace in a "detected memory leaks" section part. Its header
+    # matches the following pattern.
+    'detected memory leaks': (
+        re.compile(r'^Direct leak of \d+ byte\(s\) in \d+ object\(s\) allocated from:$'),
+    ),
+    # There are two relevant stack traces involved in one "lock-order-inversion" error section part.
+    # Both of their headers match the same pattern.
+    'lock-order-inversion': (
+        re.compile(r'^\s+Mutex M\d+ acquired here while holding mutex M\d+ in .*$'),
+        re.compile(r'^\s+Mutex M\d+ acquired here while holding mutex M\d+ in .*$'),
+    ),
+    # Remaining sanitizer errors have the only/most relevant stack trace first in a section part, so
+    # we place no restrictions on the pattern of the header. We just find the first stack trace.
+}, defaultdict(lambda: (re.compile(r'^.*$'),)))
+
+# Stack trace lines follow a "stack trace begin" line and match the following pattern.
+_FIND_STACK_TRACE_LINE_REGEX = re.compile(r'^\s*#\d+\s+.*$')
```

Comment:
If all stacktrace lines begin with some whitespace should this regexp be changed to the suggestion below to match one or more rather than zero or more whitespace characters?
Otherwise I think it's technically possible for some of the stack trace beginning regexps to overlap with this one.

```suggestion
_FIND_STACK_TRACE_LINE_REGEX = re.compile(r'^\s+#\d+\s+.*$')
```